### ISSUE TITLE
Portable mode issue

### ISSUE THREAD

**Issue by daknour** [WPN-XM/WPN-XM#384]
I have some issues with wpn-xm in its portable mode. I am using "*WPNXM-0.8.2-Full-Setup-php54-w32.exe*" and chose the full and complete installation on a usb stick. When I change the computer, there are some troubles that I think they have something to do with the disc unit. For example, lets say I install wpn-xm on unit "**G:*", then select *portable mode* and unselect shorcuts and icons. Everything works fine until I move to another computer and the assigned disc unit to the USB changes, lets say, to "**I:**" or "**J:**".

The original installation was made in Windows 7, all the files were successfully installed and the first server start registered the following **error.log** file:

> 2015/04/28 21:13:58 [error] 2404#5972: *21 WSARecv() failed (10054: FormatMessage() error:(15100)) while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "GET /tools/webinterface/index.php?page=daemon&action=restart&daemon=php HTTP/1.1", upstream: "fastcgi://127.0.0.1:9100", host: "localhost", referrer: "http://localhost/tools/webinterface/index.php?page=config"
> 2015/04/28 21:15:05 [error] 2404#5972: *25 upstream timed out (10060: FormatMessage() error:(15105)) while connecting to upstream, client: 127.0.0.1, server: localhost, request: "GET /tools/webinterface/index.php?page=config HTTP/1.1", upstream: "fastcgi://127.0.0.1:9100", host: "localhost"
> 2015/04/28 21:20:49 [notice] 6084#5960: signal process started
> 2015/04/28 21:22:29 [notice] 4904#5488: signal process started
> 2015/04/28 21:22:29 [error] 4904#5488: CreateFile() "G:/server/logs/nginx.pid" failed (2: FormatMessage() error:(15100))

Despite of that error, everything seemed to work fine in the original computer. But when I move the USB with the portable wpn-xm server the server doesn't work. The computer I am exeriencing more problems runs Windows XP, but it worked OK with older versions of portable wpn-xm server.

This is what happens:

If I start the **wpn-xm server**, there is an error on start. As in my computer it is shown in spanish, I have translated the error message to english. What I mean is that my translation may differ with the one showed by windows in english:
> "The application failed to initialize properly (0xc00000005). Click OK to terminate the application."

The program seems to manage that error enough to let me try to start the different services in the control panel. 

Then, if I try to start **Nginx**, the computer shows me this error:
> "Exception Processing Message c0000013 Parameters 75b1bf7c 4 75b1bf7c 75b1bf7c"

If I try to start **PHP**, the computer shows me the next error:
> "PHP Error. The process failed to start.
> Either the invoked program is missing, or you may have insufficient permissions to invoke the program."

With **MariaDB** I receive two errors. The first one is shown in spanish in my computer. I have translated it to english and my translation may differ with the one showed by windows in english:
> "The application failed to initialize properly (0xc00000005). Click OK to terminate the application."

And then, the second error:
> "MariaDB Error. The process crashed some time after starting succesfully"

Some other things I see that are not the cause, but may be a symptom or a clue. If the server was working on a previous computer where the USB was assigned the "**G:**" unit, and the new computer assigns the USB the "**J:**" disc unit, in the new computer, the file "`J:\server\bin\php\php.ini`" has the following variable referencing another unit, not the actual one:

`error_log =g:\server\logs\php_error.log`

The file "`J:\server\bin\mariadb\my.ini`" has a similar issue:

`log-error=g:/server/logs/mariadb_error.log`

**Comment by jakoch** [comment]
Hey Daknour,

thank you for trying out WPN-XM and reporting these issues.

Before i address the "portable mode" issue, let's address the **"control panel crash"**  first:

> If I start the wpn-xm server, there is an error on start. ....
> "The application failed to initialize properly (0xc00000005). Click OK to terminate the application."
The program seems to manage that error enough to let me try to start the different services in the control panel.

This issue was related to PostgreSQL (the Control Panel tried to start it, when it was not installed).
I fixed this in the last week and released a new version of the SCP. 

Please update your Control Panel manually to the latest version.
Just unzip the executable into the root folder of the server.
- [Server Control Panel v0.8.2 x86](https://github.com/WPN-XM/server-control-panel/releases/download/0.8.2/wpnxm-scp-x86_boxed.zip)

----

Now to the **"portable mode"** issue:

the "portable mode" has not landed, yet. The stack is not really "portable" at the moment.
I've prepared the installation wizard to provide a directive indicating the "portable mode".
If this is activated during installation, then all write operations to the registry are skipped.
But i never finished this feature - also nobody requested it, until now.
I think, the relocation is blocked solely by path issues.

When you try to "move" the stack to another drive/disc, you will run into trouble, 
because the installer writes the installation application path into the config files.
That means that the configuration files contain hardcoded paths the original installation environment,
and some servers will not start up in a new "location".
And what you described is exactly this situation. The paths are all from the old env.

In order to provide a "portable mode" a "relocation" script is needed.
This script needs to replace "all old paths" in "all config files" with "the path of the new location".
The user would have to run the relocation script once, after moving the stack to a new drive.
So, in the end, this little script will provide automatic config file adjustment.

- [ ] provide a PHP relocation script 
- [ ] batch file for script invocation, which uses an "unconfigured" PHP (because `php.ini` itself gets modified)
- [ ] identify all directives in all config files and allow modification

Contributions and ideas are welcome...

Regards, Jens

**Comment by daknour** [comment]
First of all, thanks a lot for the reply.

1.- I have replaced the file "wpn-xm.exe" with the one you provided in "Server Control Panel v0.8.2 x86", but things have not changed. If I start the server, the error message still appears. In fact, it appears twice and then the Control Panel starts:
> "The application failed to initialize properly (0xc00000005). Click OK to terminate the application."

I played with the .ini files and I discovered something interesting. I changed the drive/disc unit for the log-error in the new computer and I forgot to modify it again in the original computer. Then, in the In the original computer, MariaDB starts and stops inmediatlly. Using the right disc/drive unit solves the problem, but it just happens in the original computer where I installed the "_portable_" WPN-XM server. 

So, I am trying a second Full Installation in the same USB pendrive. This time I am doing it in another computer. Maybe this way I can use a different "_portable_" server in each computer. Could I use a www directory outside of the "server" tree?

A curious thing I have seen with several installations and different versions is that the progress bar reaches 103%. And at the end of the installation, the error message always appears:
> "The application failed to initialize properly (0xc00000005). Click OK to terminate the application."

I have tried the installation in a third computer with Windows XP. Appart form those errors, I am not able to make the server to work. Are there any compatibility issues with Windows XP?

About the scripts, it would be neccesary to find out first which files and lines are necesary to modify. I am afraid I have not found them yet. I don't know if I will be able to write those scripts, but I could try.

**Comment by jakoch** [comment]
Thank you for testing and fiddling around with this problem.
I would be happy, if we can figure out, why the "The application failed to initialize properly (0xc00000005)" error pops up.

5) Windows XP compatibility issues
The PHP 5.4 version is the last minor version supporting Windows XP (and 2003).
In other words: PHP 5.4 should work on XP.

1) When you start "wpn-xm.exe", the installed servers are detected automatically and their version numbers are fetched. This is done by executing, for instance, `php -v`. You could try to run the same command on the CLI in the PHP folder to see, if PHP starts correctly.
Try to use `php -v -n`, then the php.ini file will be ignored.
I'm not really sure, what causes this error. 
It could be missing dependency.

Question from 3) 

> Could I use a www directory outside of the "server" tree?
   
Yes, you could try to adjust the server.root directive. That means, when you have a folder like `C:\server` and `C:\server\www` and you set `server root   ../www`, Nginx will serve from `C:\www`. 
It's also possible to use absolute paths, but the stack intentionally uses relative paths.

`nginx.conf`
```
 server {
        # access from localhost only
        listen       127.0.0.1:80;
        server_name  localhost;
        root   ../www;

        ..
    }
```

4. "progress bar reaches 103%" 

Good catch. I've opened a new issue for this: https://github.com/WPN-XM/WPN-XM/issues/385

6. relocation script
>  I don't know if I will be able to write those scripts, but I could try.
I guess, we have to walk through all configuration files and create a little map, which lines or directives to modify/update.  A good indicator are all calls to SetIniString() inside the installer,
https://github.com/WPN-XM/WPN-XM/blob/master/installers/full-php54-w32.iss#L1283
I can only encourage you to contribute. The script itself isn't that hard to write, just read file, replace string, write file.. it's harder to test and get the relocation stable.

**Comment by daknour** [comment]
Thanks a lot! Now I have several things to work on. 

First of all, I suppose I should try a new installation with PHP 5.4, at least while I use the server in Windows XP.

I don't know if it could be a good idea allowing to switch between different versions of PHP with the server when we solve the portability.  

I wonder if other applications using ports 80 and 443 may be associated with this errors I observe, so I am trying to guess it in all the computers I use. And I wonder if another components (MariaDB, for example), have other compatibility issues with XP.

And I will take a look at the calls to SetIniString() inside the installer. 

Let's see what I get in the following days.

Thanks again!

**Comment by jakoch** [comment]
Blocked by #471
